File: jscomp/test/set_gen.js
'use strict';

var List = require("../../lib/js/list.js");
var Curry = require("../../lib/js/curry.js");
var Pervasives = require("../../lib/js/pervasives.js");
var Caml_exceptions = require("../../lib/js/caml_exceptions.js");

function cons_enum(_s, _e) {
  while(true) {
    var e = _e;
    var s = _s;
    if (typeof s !== "object") {
      return e;
    }
    _e = {
      TAG: "More",
      _0: s._1,
      _1: s._2,
      _2: e
    };
    _s = s._0;
    continue ;
  };
}

function height(param) {
  if (typeof param !== "object") {
    return 0;
  } else {
    return param._3;
  }
}

function min_elt(_param) {
  while(true) {
    var param = _param;
    if (typeof param !== "object") {
      throw {
            RE_EXN_ID: "Not_found",
            Error: new Error()
          };
    }
    var l = param._0;
    if (typeof l !== "object") {
      return param._1;
    }
    _param = l;
    continue ;
  };
}

function max_elt(_param) {
  while(true) {
    var param = _param;
    if (typeof param !== "object") {
      throw {
            RE_EXN_ID: "Not_found",
            Error: new Error()
          };
    }
    var r = param._2;
    if (typeof r !== "object") {
      return param._1;
    }
    _param = r;
    continue ;
  };
}

function is_empty(param) {
  if (typeof param !== "object") {
    return true;
  } else {
    return false;
  }
}

function cardinal_aux(_acc, _param) {
  while(true) {
    var param = _param;
    var acc = _acc;
    if (typeof param !== "object") {
      return acc;
    }
    _param = param._0;
    _acc = cardinal_aux(acc + 1 | 0, param._2);
    continue ;
  };
}

function cardinal(s) {
  return cardinal_aux(0, s);
}

function elements_aux(_accu, _param) {
  while(true) {
    var param = _param;
    var accu = _accu;
    if (typeof param !== "object") {
      return accu;
    }
    _param = param._0;
    _accu = {
      hd: param._1,
      tl: elements_aux(accu, param._2)
    };
    continue ;
  };
}

function elements(s) {
  return elements_aux(/* [] */0, s);
}

function iter(f, _param) {
  while(true) {
    var param = _param;
    if (typeof param !== "object") {
      return ;
    }
    iter(f, param._0);
    Curry._1(f, param._1);
    _param = param._2;
    continue ;
  };
}

function fold(f, _s, _accu) {
  while(true) {
    var accu = _accu;
    var s = _s;
    if (typeof s !== "object") {
      return accu;
    }
    _accu = Curry._2(f, s._1, fold(f, s._0, accu));
    _s = s._2;
    continue ;
  };
}

function for_all(p, _param) {
  while(true) {
    var param = _param;
    if (typeof param !== "object") {
      return true;
    }
    if (!Curry._1(p, param._1)) {
      return false;
    }
    if (!for_all(p, param._0)) {
      return false;
    }
    _param = param._2;
    continue ;
  };
}

function exists(p, _param) {
  while(true) {
    var param = _param;
    if (typeof param !== "object") {
      return false;
    }
    if (Curry._1(p, param._1)) {
      return true;
    }
    if (exists(p, param._0)) {
      return true;
    }
    _param = param._2;
    continue ;
  };
}

function max_int3(a, b, c) {
  if (a >= b) {
    if (a >= c) {
      return a;
    } else {
      return c;
    }
  } else if (b >= c) {
    return b;
  } else {
    return c;
  }
}

function max_int_2(a, b) {
  if (a >= b) {
    return a;
  } else {
    return b;
  }
}

var Height_invariant_broken = /* @__PURE__ */Caml_exceptions.create("Set_gen.Height_invariant_broken");

var Height_diff_borken = /* @__PURE__ */Caml_exceptions.create("Set_gen.Height_diff_borken");

function check_height_and_diff(param) {
  if (typeof param !== "object") {
    return 0;
  }
  var h = param._3;
  var hl = check_height_and_diff(param._0);
  var hr = check_height_and_diff(param._2);
  if (h !== (max_int_2(hl, hr) + 1 | 0)) {
    throw {
          RE_EXN_ID: Height_invariant_broken,
          Error: new Error()
        };
  }
  var diff = Pervasives.abs(hl - hr | 0);
  if (diff > 2) {
    throw {
          RE_EXN_ID: Height_diff_borken,
          Error: new Error()
        };
  }
  return h;
}

function check(tree) {
  check_height_and_diff(tree);
}

function create(l, v, r) {
  var hl;
  hl = typeof l !== "object" ? 0 : l._3;
  var hr;
  hr = typeof r !== "object" ? 0 : r._3;
  return {
          TAG: "Node",
          _0: l,
          _1: v,
          _2: r,
          _3: hl >= hr ? hl + 1 | 0 : hr + 1 | 0
        };
}

function internal_bal(l, v, r) {
  var hl;
  hl = typeof l !== "object" ? 0 : l._3;
  var hr;
  hr = typeof r !== "object" ? 0 : r._3;
  if (hl > (hr + 2 | 0)) {
    if (typeof l !== "object") {
      throw {
            RE_EXN_ID: "Assert_failure",
            _1: [
              "set_gen.ml",
              225,
              15
            ],
            Error: new Error()
          };
    }
    var lr = l._2;
    var lv = l._1;
    var ll = l._0;
    if (height(ll) >= height(lr)) {
      return create(ll, lv, create(lr, v, r));
    }
    if (typeof lr === "object") {
      return create(create(ll, lv, lr._0), lr._1, create(lr._2, v, r));
    }
    throw {
          RE_EXN_ID: "Assert_failure",
          _1: [
            "set_gen.ml",
            235,
            19
          ],
          Error: new Error()
        };
  }
  if (hr <= (hl + 2 | 0)) {
    return {
            TAG: "Node",
            _0: l,
            _1: v,
            _2: r,
            _3: hl >= hr ? hl + 1 | 0 : hr + 1 | 0
          };
  }
  if (typeof r !== "object") {
    throw {
          RE_EXN_ID: "Assert_failure",
          _1: [
            "set_gen.ml",
            245,
            15
          ],
          Error: new Error()
        };
  }
  var rr = r._2;
  var rv = r._1;
  var rl = r._0;
  if (height(rr) >= height(rl)) {
    return create(create(l, v, rl), rv, rr);
  }
  if (typeof rl === "object") {
    return create(create(l, v, rl._0), rl._1, create(rl._2, rv, rr));
  }
  throw {
        RE_EXN_ID: "Assert_failure",
        _1: [
          "set_gen.ml",
          251,
          19
        ],
        Error: new Error()
      };
}

function remove_min_elt(param) {
  if (typeof param !== "object") {
    throw {
          RE_EXN_ID: "Invalid_argument",
          _1: "Set.remove_min_elt",
          Error: new Error()
        };
  }
  var l = param._0;
  if (typeof l !== "object") {
    return param._2;
  } else {
    return internal_bal(remove_min_elt(l), param._1, param._2);
  }
}

function singleton(x) {
  return {
          TAG: "Node",
          _0: "Empty",
          _1: x,
          _2: "Empty",
          _3: 1
        };
}

function internal_merge(l, r) {
  if (typeof l !== "object") {
    return r;
  } else if (typeof r !== "object") {
    return l;
  } else {
    return internal_bal(l, min_elt(r), remove_min_elt(r));
  }
}

function add_min_element(v, param) {
  if (typeof param !== "object") {
    return singleton(v);
  } else {
    return internal_bal(add_min_element(v, param._0), param._1, param._2);
  }
}

function add_max_element(v, param) {
  if (typeof param !== "object") {
    return singleton(v);
  } else {
    return internal_bal(param._0, param._1, add_max_element(v, param._2));
  }
}

function internal_join(l, v, r) {
  if (typeof l !== "object") {
    return add_min_element(v, r);
  }
  var lh = l._3;
  if (typeof r !== "object") {
    return add_max_element(v, l);
  }
  var rh = r._3;
  if (lh > (rh + 2 | 0)) {
    return internal_bal(l._0, l._1, internal_join(l._2, v, r));
  } else if (rh > (lh + 2 | 0)) {
    return internal_bal(internal_join(l, v, r._0), r._1, r._2);
  } else {
    return create(l, v, r);
  }
}

function internal_concat(t1, t2) {
  if (typeof t1 !== "object") {
    return t2;
  } else if (typeof t2 !== "object") {
    return t1;
  } else {
    return internal_join(t1, min_elt(t2), remove_min_elt(t2));
  }
}

function filter(p, param) {
  if (typeof param !== "object") {
    return "Empty";
  }
  var v = param._1;
  var l$p = filter(p, param._0);
  var pv = Curry._1(p, v);
  var r$p = filter(p, param._2);
  if (pv) {
    return internal_join(l$p, v, r$p);
  } else {
    return internal_concat(l$p, r$p);
  }
}

function partition(p, param) {
  if (typeof param !== "object") {
    return [
            "Empty",
            "Empty"
          ];
  }
  var v = param._1;
  var match = partition(p, param._0);
  var lf = match[1];
  var lt = match[0];
  var pv = Curry._1(p, v);
  var match$1 = partition(p, param._2);
  var rf = match$1[1];
  var rt = match$1[0];
  if (pv) {
    return [
            internal_join(lt, v, rt),
            internal_concat(lf, rf)
          ];
  } else {
    return [
            internal_concat(lt, rt),
            internal_join(lf, v, rf)
          ];
  }
}

function of_sorted_list(l) {
  var sub = function (n, l) {
    switch (n) {
      case 0 :
          return [
                  "Empty",
                  l
                ];
      case 1 :
          if (l) {
            return [
                    {
                      TAG: "Node",
                      _0: "Empty",
                      _1: l.hd,
                      _2: "Empty",
                      _3: 1
                    },
                    l.tl
                  ];
          }
          break;
      case 2 :
          if (l) {
            var match = l.tl;
            if (match) {
              return [
                      {
                        TAG: "Node",
                        _0: {
                          TAG: "Node",
                          _0: "Empty",
                          _1: l.hd,
                          _2: "Empty",
                          _3: 1
                        },
                        _1: match.hd,
                        _2: "Empty",
                        _3: 2
                      },
                      match.tl
                    ];
            }
            
          }
          break;
      case 3 :
          if (l) {
            var match$1 = l.tl;
            if (match$1) {
              var match$2 = match$1.tl;
              if (match$2) {
                return [
                        {
                          TAG: "Node",
                          _0: {
                            TAG: "Node",
                            _0: "Empty",
                            _1: l.hd,
                            _2: "Empty",
                            _3: 1
                          },
                          _1: match$1.hd,
                          _2: {
                            TAG: "Node",
                            _0: "Empty",
                            _1: match$2.hd,
                            _2: "Empty",
                            _3: 1
                          },
                          _3: 2
                        },
                        match$2.tl
                      ];
              }
              
            }
            
          }
          break;
      default:
        
    }
    var nl = n / 2 | 0;
    var match$3 = sub(nl, l);
    var l$1 = match$3[1];
    if (l$1) {
      var match$4 = sub((n - nl | 0) - 1 | 0, l$1.tl);
      return [
              create(match$3[0], l$1.hd, match$4[0]),
              match$4[1]
            ];
    }
    throw {
          RE_EXN_ID: "Assert_failure",
          _1: [
            "set_gen.ml",
            361,
            14
          ],
          Error: new Error()
        };
  };
  return sub(List.length(l), l)[0];
}

function of_sorted_array(l) {
  var sub = function (start, n, l) {
    if (n === 0) {
      return "Empty";
    }
    if (n === 1) {
      var x0 = l[start];
      return {
              TAG: "Node",
              _0: "Empty",
              _1: x0,
              _2: "Empty",
              _3: 1
            };
    }
    if (n === 2) {
      var x0$1 = l[start];
      var x1 = l[start + 1 | 0];
      return {
              TAG: "Node",
              _0: {
                TAG: "Node",
                _0: "Empty",
                _1: x0$1,
                _2: "Empty",
                _3: 1
              },
              _1: x1,
              _2: "Empty",
              _3: 2
            };
    }
    if (n === 3) {
      var x0$2 = l[start];
      var x1$1 = l[start + 1 | 0];
      var x2 = l[start + 2 | 0];
      return {
              TAG: "Node",
              _0: {
                TAG: "Node",
                _0: "Empty",
                _1: x0$2,
                _2: "Empty",
                _3: 1
              },
              _1: x1$1,
              _2: {
                TAG: "Node",
                _0: "Empty",
                _1: x2,
                _2: "Empty",
                _3: 1
              },
              _3: 2
            };
    }
    var nl = n / 2 | 0;
    var left = sub(start, nl, l);
    var mid = start + nl | 0;
    var v = l[mid];
    var right = sub(mid + 1 | 0, (n - nl | 0) - 1 | 0, l);
    return create(left, v, right);
  };
  return sub(0, l.length, l);
}

function is_ordered(cmp, tree) {
  var is_ordered_min_max = function (tree) {
    if (typeof tree !== "object") {
      return "Empty";
    }
    var r = tree._2;
    var v = tree._1;
    var match = is_ordered_min_max(tree._0);
    if (typeof match === "object") {
      var match$1 = match.VAL;
      var max_v = match$1[1];
      var min_v = match$1[0];
      var match$2 = is_ordered_min_max(r);
      if (typeof match$2 !== "object") {
        if (match$2 === "Empty" && Curry._2(cmp, max_v, v) < 0) {
          return {
                  NAME: "V",
                  VAL: [
                    min_v,
                    v
                  ]
                };
        } else {
          return "No";
        }
      }
      var match$3 = match$2.VAL;
      if (Curry._2(cmp, max_v, match$3[0]) < 0) {
        return {
                NAME: "V",
                VAL: [
                  min_v,
                  match$3[1]
                ]
              };
      } else {
        return "No";
      }
    }
    if (match !== "Empty") {
      return "No";
    }
    var match$4 = is_ordered_min_max(r);
    if (typeof match$4 !== "object") {
      if (match$4 === "Empty") {
        return {
                NAME: "V",
                VAL: [
                  v,
                  v
                ]
              };
      } else {
        return "No";
      }
    }
    var match$5 = match$4.VAL;
    if (Curry._2(cmp, v, match$5[0]) < 0) {
      return {
              NAME: "V",
              VAL: [
                v,
                match$5[1]
              ]
            };
    } else {
      return "No";
    }
  };
  return is_ordered_min_max(tree) !== "No";
}

function invariant(cmp, t) {
  check_height_and_diff(t);
  return is_ordered(cmp, t);
}

function compare_aux(cmp, _e1, _e2) {
  while(true) {
    var e2 = _e2;
    var e1 = _e1;
    if (typeof e1 !== "object") {
      if (typeof e2 !== "object") {
        return 0;
      } else {
        return -1;
      }
    }
    if (typeof e2 !== "object") {
      return 1;
    }
    var c = Curry._2(cmp, e1._0, e2._0);
    if (c !== 0) {
      return c;
    }
    _e2 = cons_enum(e2._1, e2._2);
    _e1 = cons_enum(e1._1, e1._2);
    continue ;
  };
}

function compare(cmp, s1, s2) {
  return compare_aux(cmp, cons_enum(s1, "End"), cons_enum(s2, "End"));
}

var empty = "Empty";

var choose = min_elt;

exports.cons_enum = cons_enum;
exports.height = height;
exports.min_elt = min_elt;
exports.max_elt = max_elt;
exports.empty = empty;
exports.is_empty = is_empty;
exports.cardinal_aux = cardinal_aux;
exports.cardinal = cardinal;
exports.elements_aux = elements_aux;
exports.elements = elements;
exports.choose = choose;
exports.iter = iter;
exports.fold = fold;
exports.for_all = for_all;
exports.exists = exists;
exports.max_int3 = max_int3;
exports.max_int_2 = max_int_2;
exports.Height_invariant_broken = Height_invariant_broken;
exports.Height_diff_borken = Height_diff_borken;
exports.check_height_and_diff = check_height_and_diff;
exports.check = check;
exports.create = create;
exports.internal_bal = internal_bal;
exports.remove_min_elt = remove_min_elt;
exports.singleton = singleton;
exports.internal_merge = internal_merge;
exports.add_min_element = add_min_element;
exports.add_max_element = add_max_element;
exports.internal_join = internal_join;
exports.internal_concat = internal_concat;
exports.filter = filter;
exports.partition = partition;
exports.of_sorted_list = of_sorted_list;
exports.of_sorted_array = of_sorted_array;
exports.is_ordered = is_ordered;
exports.invariant = invariant;
exports.compare_aux = compare_aux;
exports.compare = compare;
/* No side effect */

File: jscomp/ext/set_gen.ml
(***********************************************************************)
(*                                                                     *)
(*                                OCaml                                *)
(*                                                                     *)
(*            Xavier Leroy, projet Cristal, INRIA Rocquencourt         *)
(*                                                                     *)
(*  Copyright 1996 Institut National de Recherche en Informatique et   *)
(*  en Automatique.  All rights reserved.  This file is distributed    *)
(*  under the terms of the GNU Library General Public License, with    *)
(*  the special exception on linking described in file ../LICENSE.     *)
(*                                                                     *)
(***********************************************************************)
[@@@warnerror "+55"]
(** balanced tree based on stdlib distribution *)

type 'a t0 = 
  | Empty 
  | Leaf of  'a 
  | Node of { l : 'a t0 ; v :  'a ; r : 'a t0 ; h :  int }

let empty = Empty
let  [@inline] height = function
  | Empty -> 0 
  | Leaf _ -> 1
  | Node {h} -> h   

let [@inline] calc_height a b =  
  (if a >= b then a else b) + 1 

(* 
    Invariants: 
    1. {[ l < v < r]}
    2. l and r balanced 
    3. [height l] - [height r] <= 2
*)
let [@inline] unsafe_node v l  r h = 
  Node{l;v;r; h }         

let [@inline] unsafe_node_maybe_leaf v l r h =   
  if h = 1 then Leaf v   
  else Node{l;v;r; h }         

let [@inline] singleton x = Leaf x

let [@inline] unsafe_two_elements x v = 
  unsafe_node v (singleton x) empty 2 
  
type 'a t = 'a t0 = private
  | Empty 
  | Leaf of 'a
  | Node of { l : 'a t0 ; v :  'a ; r : 'a t0 ; h :  int }


(* Smallest and greatest element of a set *)

let rec min_elt = function
  | Empty -> raise Not_found
  | Leaf v -> v 
  | Node{l; v} ->
    match l with 
    | Empty -> v 
    | Leaf _
    | Node _ ->  min_elt l

let rec max_elt = function
  | Empty -> raise Not_found
  | Leaf v -> v 
  | Node{ v; r} -> 
    match r with 
    | Empty -> v 
    | Leaf _
    | Node _ -> max_elt r





let is_empty = function Empty -> true | _ -> false

let rec cardinal_aux acc  = function
  | Empty -> acc 
  | Leaf _ -> acc + 1
  | Node {l;r} -> 
    cardinal_aux  (cardinal_aux (acc + 1)  r ) l 

let cardinal s = cardinal_aux 0 s 

let rec elements_aux accu = function
  | Empty -> accu
  | Leaf v -> v :: accu
  | Node{l; v; r} -> elements_aux (v :: elements_aux accu r) l

let elements s =
  elements_aux [] s

let choose = min_elt

let rec iter  x f = match x with
  | Empty -> ()
  | Leaf v -> f v 
  | Node {l; v; r} -> iter l f ; f v; iter r f 

let rec fold s accu f =
  match s with
  | Empty -> accu
  | Leaf v -> f v accu
  | Node{l; v; r} -> fold r (f v (fold l accu f)) f 

let rec for_all x p = match x with
  | Empty -> true
  | Leaf v -> p v 
  | Node{l; v; r} -> p v && for_all l p && for_all r p 

let rec exists x p = match x with
  | Empty -> false
  | Leaf v -> p v 
  | Node {l; v; r} -> p v || exists l p  || exists r p





exception Height_invariant_broken
exception Height_diff_borken 

let rec check_height_and_diff = 
  function 
  | Empty -> 0
  | Leaf _ -> 1
  | Node{l;r;h} -> 
    let hl = check_height_and_diff l in
    let hr = check_height_and_diff r in
    if h <>  calc_height hl hr  then raise Height_invariant_broken
    else  
      let diff = (abs (hl - hr)) in  
      if  diff > 2 then raise Height_diff_borken 
      else h     

let check tree = 
  ignore (check_height_and_diff tree)

(* Same as create, but performs one step of rebalancing if necessary.
    Invariants:
    1. {[ l < v < r ]}
    2. l and r balanced 
    3. | height l - height r | <= 3.

    Proof by indunction

    Lemma: the height of  [bal l v r] will bounded by [max l r] + 1 
*)
let internal_bal l v r : _ t =
  let hl = height l in
  let hr = height r in
  if hl > hr + 2 then 
    let [@warning "-8"] Node ({l=ll;r= lr} as l) = l in 
    let hll = height ll in 
    let hlr = height lr in 
    if hll >= hlr then
      let hnode = calc_height hlr hr in       
      unsafe_node l.v 
        ll  
        (unsafe_node_maybe_leaf v lr  r hnode ) 
        (calc_height hll hnode)
    else       
      let [@warning "-8"] Node ({l = lrl; r = lrr } as lr) = lr in 
      let hlrl = height lrl in 
      let hlrr = height lrr in 
      let hlnode = calc_height hll hlrl in 
      let hrnode = calc_height hlrr hr in 
      unsafe_node lr.v 
        (unsafe_node_maybe_leaf l.v ll  lrl hlnode)  
        (unsafe_node_maybe_leaf v lrr  r hrnode)
        (calc_height hlnode hrnode)
  else if hr > hl + 2 then begin    
    let [@warning "-8"] Node ({l=rl; r=rr} as r) = r in 
    let hrr = height rr in 
    let hrl = height rl in 
    if hrr >= hrl then
      let hnode = calc_height hl hrl in
      unsafe_node r.v 
        (unsafe_node_maybe_leaf v l  rl hnode) 
        rr 
        (calc_height hnode hrr )
    else begin
      let [@warning "-8"] Node ({l = rll ; r = rlr } as rl) = rl in 
      let hrll = height rll in 
      let hrlr = height rlr in 
      let hlnode = (calc_height hl hrll) in
      let hrnode = (calc_height hrlr hrr) in
      unsafe_node rl.v 
        (unsafe_node_maybe_leaf v l rll hlnode)  
        (unsafe_node_maybe_leaf r.v rlr rr hrnode)
        (calc_height hlnode hrnode)
    end
  end else
    unsafe_node_maybe_leaf v l  r (calc_height hl hr)


let rec remove_min_elt = function
    Empty -> invalid_arg "Set.remove_min_elt"
  | Leaf _ -> empty  
  | Node{l=Empty; r} -> r
  | Node{l; v; r} -> internal_bal (remove_min_elt l) v r



(* 
   All elements of l must precede the elements of r.
       Assume | height l - height r | <= 2.
   weak form of [concat] 
*)

let internal_merge l r =
  match (l, r) with
  | (Empty, t) -> t
  | (t, Empty) -> t
  | (_, _) -> internal_bal l (min_elt r) (remove_min_elt r)


(* Beware: those two functions assume that the added v is *strictly*
    smaller (or bigger) than all the present elements in the tree; it
    does not test for equality with the current min (or max) element.
    Indeed, they are only used during the "join" operation which
    respects this precondition.
*)

let rec add_min_element v = function
  | Empty -> singleton v
  | Leaf x -> unsafe_two_elements v x
  | Node {l; v=x; r} ->
    internal_bal (add_min_element v l) x r

let rec add_max_element v = function
  | Empty -> singleton v
  | Leaf x -> unsafe_two_elements x v
  | Node {l; v=x; r} ->
    internal_bal l x (add_max_element v r)

(** 
    Invariants:
    1. l < v < r 
    2. l and r are balanced 

    Proof by induction
    The height of output will be ~~ (max (height l) (height r) + 2)
    Also use the lemma from [bal]
*)
let rec internal_join l v r =
  match (l, r) with
    (Empty, _) -> add_min_element v r
  | (_, Empty) -> add_max_element v l
  | Leaf lv, Node {h = rh} ->
    if rh > 3 then 
      add_min_element lv (add_min_element v r ) (* FIXME: could inlined *)
    else unsafe_node  v l r (rh + 1)
  | Leaf _, Leaf _ -> 
    unsafe_node  v l r 2
  | Node {h = lh}, Leaf rv ->
    if lh > 3 then       
      add_max_element rv (add_max_element v l)
    else unsafe_node  v l r (lh + 1)    
  | (Node{l=ll;v= lv;r= lr;h= lh}, Node {l=rl; v=rv; r=rr; h=rh}) ->
    if lh > rh + 2 then 
      (* proof by induction:
         now [height of ll] is [lh - 1] 
      *)
      internal_bal ll lv (internal_join lr v r) 
    else
    if rh > lh + 2 then internal_bal (internal_join l v rl) rv rr 
    else unsafe_node  v l r (calc_height lh rh)


(*
    Required Invariants: 
    [t1] < [t2]  
*)
let internal_concat t1 t2 =
  match (t1, t2) with
  | (Empty, t) -> t
  | (t, Empty) -> t
  | (_, _) -> internal_join t1 (min_elt t2) (remove_min_elt t2)

(* let rec filter x p = match x with 
  | Empty -> Empty
  | Node {l; v; r} ->
    (* call [p] in the expected left-to-right order *)
    let l' = filter l p in
    let pv = p v in
    let r' = filter r p in
    if pv then internal_join l' v r' else internal_concat l' r'
 *)

let rec partition x p = match x with 
  | Empty -> (empty, empty)
  | Leaf v -> let pv = p v in if pv then x, empty else empty, x
  | Node{l; v; r} ->
    (* call [p] in the expected left-to-right order *)
    let (lt, lf) = partition l p in
    let pv = p v in
    let (rt, rf) = partition r p in
    if pv
    then (internal_join lt v rt, internal_concat lf rf)
    else (internal_concat lt rt, internal_join lf v rf)


let of_sorted_array l =   
  let rec sub start n l  =
    if n = 0 then empty else 
    if n = 1 then 
      let x0 = Array.unsafe_get l start in
      singleton x0
    else if n = 2 then     
      let x0 = Array.unsafe_get l start in 
      let x1 = Array.unsafe_get l (start + 1) in 
      unsafe_node x1 (singleton x0)  empty 2 else
    if n = 3 then 
      let x0 = Array.unsafe_get l start in 
      let x1 = Array.unsafe_get l (start + 1) in
      let x2 = Array.unsafe_get l (start + 2) in
      unsafe_node x1 (singleton x0)  (singleton x2) 2
    else 
      let nl = n / 2 in
      let left = sub start nl l in
      let mid = start + nl in 
      let v = Array.unsafe_get l mid in 
      let right = sub (mid + 1) (n - nl - 1) l in        
      unsafe_node v left  right (calc_height (height left) (height right))
  in
  sub 0 (Array.length l) l 

let is_ordered ~cmp tree =
  let rec is_ordered_min_max tree =
    match tree with
    | Empty -> `Empty
    | Leaf v -> `V (v,v)
    | Node {l;v;r} -> 
      begin match is_ordered_min_max l with
        | `No -> `No 
        | `Empty ->
          begin match is_ordered_min_max r with
            | `No  -> `No
            | `Empty -> `V (v,v)
            | `V(l,r) ->
              if cmp v l < 0 then
                `V(v,r)
              else
                `No
          end
        | `V(min_v,max_v)->
          begin match is_ordered_min_max r with
            | `No -> `No
            | `Empty -> 
              if cmp max_v v < 0 then 
                `V(min_v,v)
              else
                `No 
            | `V(min_v_r, max_v_r) ->
              if cmp max_v min_v_r < 0 then
                `V(min_v,max_v_r)
              else `No
          end
      end  in 
  is_ordered_min_max tree <> `No 

let invariant ~cmp t = 
  check t ; 
  is_ordered ~cmp t 

(* let rec compare_aux ~cmp e1 e2 =
  match (e1, e2) with
    (End, End) -> 0
  | (End, _)  -> -1
  | (_, End) -> 1
  | (More(v1, r1, e1), More(v2, r2, e2)) ->
    let c = cmp v1 v2 in
    if c <> 0
    then c
    else compare_aux ~cmp (cons_enum r1 e1) (cons_enum r2 e2)

let compare ~cmp s1 s2 =
  compare_aux ~cmp (cons_enum s1 End) (cons_enum s2 End) *)


module type S = sig
  type elt 
  type t
  val empty: t
  val is_empty: t -> bool
  val iter: t ->  (elt -> unit) -> unit
  val fold: t -> 'a -> (elt -> 'a -> 'a) -> 'a
  val for_all: t -> (elt -> bool) ->  bool
  val exists: t -> (elt -> bool) -> bool
  val singleton: elt -> t
  val cardinal: t -> int
  val elements: t -> elt list
  val min_elt: t -> elt
  val choose: t -> elt
  val mem: t -> elt -> bool
  val add: t -> elt -> t
  val remove: t -> elt -> t
  val union: t -> t -> t
  val inter: t -> t -> t
  val diff: t -> t -> t
  val of_list: elt list -> t
  val of_sorted_array : elt array -> t 
  val invariant : t -> bool 
  val print : Format.formatter -> t -> unit 
end 
